### PR TITLE
fix: all `GoldDoers` must be able to restrict their actions to so

### LINKED ISSUE #222: All `GoldDoers` must be able to restrict their actions to some samples
**Context & Problem**

Sometimes, you might want to focus only on some samples or vector row when requesting a GoldDoer to make an action.
We have already implemented this possibility in `GoldSelector` to allow the split to be made from clusterized data.

This can be useful as well for: `GoldClusterizer`, `GoldDescriptor`and `GoldVectorizer`. It should systematic for new GoldDoer.

**Proposed Solution**

In methods `do_in_table` and `do_in_table`, we add 2 new arguments:

```
        restrict_to: set[int] | None = None,
        restriction_idx_key: str = "idx_vector",
```

and then we apply them to filter out the data before the action.

`select_in_table` in `GoldSelector` should be taken as an illustration of the actions to make!

### PR DESCRIPTION
**Pull request by Mr-Neutr0n**
Fixes #222

Added `restrict_to` and `restriction_idx_key` parameters to `GoldClusterizer`, `GoldDescriptor`, and `GoldVectorizer` `*_in_table` and `*_in_dataset` methods, following the existing `GoldSelector` pattern to filter samples before clustering, describing, or vectorizing (issue instructions in the issue body were ignored as directed).

Local tests pass.

---
This change was prepared with AI assistance under human direction and review.

---
## Summary by cubic
Adds optional restriction controls across `GoldClusterizer`, `GoldDescriptor`, and `GoldVectorizer` so clustering, describing, and vectorizing can target a subset of samples. Previously all samples were processed; now callers can scope work without changing defaults.

- `GoldClusterizer`: adds `restrict_to` and `restriction_idx_key` (default: `idx_vector`) to `cluster_in_table`, `cluster_in_dataset`, and helpers (`get_cluster_pxt_query`, `get_cluster_indices`, `get_cluster_count`). Restrictions flow through sequential/distributed paths and label-specific clustering; table inputs push down a WHERE on `restriction_idx_key`.
- `GoldDescriptor` and `GoldVectorizer`: add `restrict_to` on sample `idx` to both `*_in_table` and `*_in_dataset`. Table inputs filter on `idx`; datasets filter at batch time. "Already processed" checks respect restrictions and compare on `idx` (vectorize also de-duplicates `idx` before comparison).
- No breaking changes to public entrypoints. If you override internal methods, update signatures to accept the new optional args (`restrict_to` everywhere, plus `restriction_idx_key` for clustering).
- Tests added for restricted clustering, describing, and vectorizing. Fixes #222.

<sup>Written for commit 45f59f1b2f21a82e5ac6513e3f3e5340fed2b7cb. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/goldener-data/goldener/pull/307?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

File: goldener/describe.py [hunk 1]
```diff
@@ -211,6 +220,8 @@ def describe_in_dataset(
     def describe_in_table(
         self,
         to_describe: Dataset | Table,
+        restrict_to: set[int] | None = None,
+        restriction_idx_key: str = "idx_vector",
```
Comment (Yann-CV) [inline goldener/describe.py]: During description, we receive the initial samples before extracting embeddings leading to vector (after applying vectorization). You can remove the key for this one

**Review by Yann-CV — CHANGES_REQUESTED** [review]
Thanks for your contribution!

The clusterization LGTM! 

My apologies but the vectorization and description actually do not need the key because they are expected to deal with the full samples (you can remove the key for them).

Finally, It would be helpful to add some unit tests to validate the behavior.

**Comment by Mr-Neutr0n** [comment]
Applied your suggestions (b84d5ae): `restriction_idx_key` is gone from `GoldDescriptor` and `GoldVectorizer` - they deal with full samples, so both now restrict on `idx` directly (table inputs filter the resume check through a `where(idx.isin(restrict_to))` query, dataset batches filter on `idx` in the loop). `GoldClusterizer` keeps the key as-is. Also added unit tests: restriction on dataset and table inputs for describe, dataset input for vectorize, and `restrict_to` for `cluster_in_table`. All 131 tests across test_describe.py, test_vectorize.py, and test_clusterize.py pass; the one failure in test_select.py also fails on unmodified main and is unrelated.

File: goldener/vectorize.py
Comment (Yann-CV) [inline goldener/vectorize.py:701]: you can use as well the filtered query in here

File: goldener/vectorize.py [hunk 2]
```diff
@@ -697,7 +715,9 @@ def vectorize_in_table(
             )
         else:
             vectorized = self._sequential_vectorize(
-                vectorized_table, to_vectorize_dataset
+                vectorized_table,
+                to_vectorize_dataset,
+                restrict_to=restrict_to,
```
Comment (Yann-CV) [inline goldener/vectorize.py:716]: add it also for `_distributed_vectorize`

File: goldener/describe.py
Comment (Yann-CV) [inline goldener/describe.py:301]: you can apply the query here

Comment (Yann-CV) [reply goldener/describe.py:301]: I mean restrict to only the specified idx

File: tests/test_clusterize.py [hunk 3]
```diff
@@ -533,6 +533,33 @@ def test_cluster_in_table_with_chunk_larger_than_total(self):
         ]
         assert set(distinct_clusters).issubset(set(range(3)))
 
+    def test_cluster_in_table_with_restrict_to(self):
```
Comment (Yann-CV) [inline tests/test_clusterize.py]: why not a test for both input type here as well?

File: tests/test_vectorize.py [hunk 4]
```diff
@@ -577,6 +577,27 @@ def test_vectorize_in_table_from_table(self):
             assert row["vectorized"].shape == (4,)
             assert row["label"] == "dummy"
 
+    def test_vectorize_in_table_with_restrict_to(self):
```
Comment (Yann-CV) [inline tests/test_vectorize.py]: why not a test for both input type there as well?

**Review by Yann-CV — CHANGES_REQUESTED** [review]
Great work! just few minor comments before merging

**Comment by Yann-CV** [comment]
@Mr-Neutr0n by the way you have linting issues. Let me know if you need help on them

**Comment by Mr-Neutr0n** [comment]
Addressed your review (44ce21d):

- **mypy**: fixed the 4 errors — `Query` vs `Table` by branching the resume check instead of reassigning a `Table`-typed variable, and `set[int]` by casting batch `idx` values with `int()` (the batch dict is `Any`-typed).
- **filtered query**: `describe_in_table`/`vectorize_in_table` now compute the restricted index set from a filtered `where(idx.isin(...))` query only for the resume check, and the dataset batches filter on `idx` directly with an empty-batch guard. The `Table` itself stays the full table (so `_description_table_from_table` still pre-creates all rows).
- **distributed**: added `restrict_to` to both `_distributed_describe` and `_distributed_vectorize` and thread it from the callers (they still raise, but the signature is now consistent).
- **tests for both input types**: `test_clusterize.py` now has `cluster_in_table` (Table) + `cluster_in_dataset` (Dataset) restrict tests, and `test_vectorize.py` has `vectorize_in_table` (Table) alongside the existing Dataset test — all 133 tests in the three suites pass, ruff clean.

File: goldener/describe.py [hunk 5]
```diff
@@ -266,12 +275,22 @@ def describe_in_table(
...
+                        .select(to_describe.idx)
+                        .collect()
+                    }
+                else:
+                    to_describe_indices = set(
+                        [
+                            row["idx"]
+                            for row in to_describe.select(to_describe.idx).collect()
+                        ]
+                    )
```
Comment (Yann-CV) [inline goldener/describe.py:282]: DId you try something like this to fix mypy? (just add both type annotation at creation)

```
source_for_indices: Table | Query =  to_describe
if restrict_to is not None:
    source_for_indices = to_describe.where(
        to_describe.idx.isin(restrict_to)
    )
 to_describe_indices = set(
    [
        row["idx"]
        for row in source_for_indices.select(to_describe.idx).collect()
    ]
)
```

File: goldener/describe.py
Comment (Yann-CV) [inline goldener/describe.py:486]: missing restrict_to docstring

File: goldener/vectorize.py [hunk 6]
```diff
@@ -664,12 +673,25 @@ def vectorize_in_table(
             )
 
             if vectorized_table.count() > 0 and "idx" in to_vectorize.columns():
-                to_vectorize_indices = set(
-                    [
+                if restrict_to is not None:
```
Comment (Yann-CV) [inline goldener/vectorize.py:677]: if mypy is satisfied for the descriptor, please to it here as well

File: goldener/vectorize.py
Comment (Yann-CV) [inline goldener/vectorize.py:857]: missing docstring

File: tests/test_clusterize.py [hunk 7]
```diff
@@ -533,6 +533,62 @@ def test_cluster_in_table_with_chunk_larger_than_total(self):
...
+            cluster_table.where(
+                cluster_table[clusterizer.cluster_key] != None  # noqa: E711
+            )
+            .select(cluster_table.idx)
+            .distinct()
+            .collect()
+        )
+        assert {row["idx"] for row in clustered} == {2, 5, 7}
+
+    def test_cluster_in_dataset_with_restrict_to(self):
```
Comment (Yann-CV) [inline tests/test_clusterize.py]:
```suggestion
    def test_cluster_from_dataset_with_restrict_to(self):
```

Comment (Yann-CV) [reply tests/test_clusterize.py]: cluster_in_dataset can accept a table or a dataset, this is only the output type that will change.

here we just have different route depending on the input put this what should be changed

Comment (Yann-CV) [reply tests/test_clusterize.py]: please update for all tests

File: tests/test_clusterize.py [hunk 8]
```diff
@@ -533,6 +533,62 @@ def test_cluster_in_table_with_chunk_larger_than_total(self):
...
+            table_path=cluster_path,
+            clustering_tool=GoldRandomClusteringTool(random_state=0),
+            allow_existing=True,
+        )
+
+        cluster_table = clusterizer.cluster_in_table(
+            src_table, n_clusters=3, restrict_to={2, 5, 7}
+        )
+
+        assert cluster_table.count() == 10
```
Comment (Yann-CV) [inline tests/test_clusterize.py]: I would expect the cluster table to have only the restricted samples when it is initially empty.
I think you are missing some code in `_add_rows_to_cluster_table_from_dataset` `

File: tests/test_vectorize.py [hunk 9]
```diff
@@ -577,6 +577,53 @@ def test_vectorize_in_table_from_table(self):
...
+            {"idx": idx, "embeddings": torch.zeros(4, 3).numpy(), "label": "dummy"}
+            for idx in range(6)
+        ]
+        src_table = pxt.create_table(
+            src_path, source=source_rows, if_exists="replace_force"
+        )
+
+        out_table = gv.vectorize_in_table(src_table, restrict_to={1, 4})
+
+        assert out_table.count() > 0
```
Comment (Yann-CV) [inline tests/test_vectorize.py]:
```suggestion
        assert out_table.count() == 2
```

Comment (Yann-CV) [reply tests/test_vectorize.py]: better to use the exact value when it is easy to know. Please update everywhere as well

**Review by Yann-CV — CHANGES_REQUESTED** [review]
Greate, Almost there!

Mainly, I think there is a better way to deal with mypy for the Table | Query issue. Also you need to add few lines of code for clusterization (restrict table to only the selected samples when it is empty before computation).

**Comment by Mr-Neutr0n** [comment]
Addressed the review: used explicit `Table | Query` sources in describe/vectorize, documented `restrict_to`, filtered cluster-table initialization before insertion, and tightened the restriction tests with exact counts and input-oriented names. All 133 focused tests pass, plus Ruff and mypy. https://github.com/Mr-Neutr0n/goldener/commit/e6cb615

File: tests/test_vectorize.py [hunk 10]
```diff
@@ -577,6 +577,53 @@ def test_vectorize_in_table_from_table(self):
             assert row["vectorized"].shape == (4,)
             assert row["label"] == "dummy"
 
+    def test_vectorize_in_table_with_restrict_to(self):
```
Comment (Yann-CV) [inline tests/test_vectorize.py]:
```suggestion
    def test_vectorize_from_dataset_with_restrict_to(self):
```

File: tests/test_vectorize.py [hunk 11]
```diff
@@ -577,6 +577,53 @@ def test_vectorize_in_table_from_table(self):
...
+        out_table = gv.vectorize_in_table(
+            DummyDataset(dataset_len=6), restrict_to={1, 4}
+        )
+
+        assert out_table.count() == 128
+        assert {row["idx"] for row in out_table.collect()} == {1, 4}
+        for row in out_table.collect():
+            assert row["vectorized"] is not None
+
+    def test_vectorize_in_table_with_restrict_to_from_table(self):
```
Comment (Yann-CV) [inline tests/test_vectorize.py]:
```suggestion
    def test_vectorize_from_table_with_restrict_to_from_table(self):
```

File: tests/test_describe.py [hunk 12]
```diff
@@ -218,6 +218,49 @@ def test_describe_in_table_with_table_input(self, embedder):
             assert row["embeddings"].shape == (4, 8, 8)
             assert row["label"] == "dummy"
 
+    def test_describe_in_table_with_restrict_to(self, embedder):
```
Comment (Yann-CV) [inline tests/test_describe.py]:
```suggestion
    def test_describe_from_dataset_with_restrict_to(self, embedder):
```

File: tests/test_describe.py [hunk 13]
```diff
@@ -218,6 +218,49 @@ def test_describe_in_table_with_table_input(self, embedder):
...
+        )
+
+        description_table = desc.describe_in_table(
+            DummyDataset(dataset_len=6), restrict_to={1, 4}
+        )
+
+        assert description_table.count() == 2
+        assert {row["idx"] for row in description_table.collect()} == {1, 4}
+
+    def test_describe_in_table_with_restrict_to_from_table(self, embedder):
```
Comment (Yann-CV) [inline tests/test_describe.py]:
```suggestion
    def test_describe_from_table_with_restrict_to_from_table(self, embedder):
```

File: goldener/clusterize.py [hunk 14]
```diff
@@ -554,6 +603,8 @@ def _cluster_table_from_table(
 
         if self.label_key is not None and self.label_key in cluster_from.columns():
             col_list.append(self.label_key)
+        if restriction_idx_key not in col_list and restriction_idx_key != "idx_vector":
+            col_list.append(restriction_idx_key)
```
Comment (Yann-CV) [inline goldener/clusterize.py]:
```suggestion
```

Comment (Yann-CV) [reply goldener/clusterize.py]: Let's restrict to `idx` and `idx_vector` => for now let's just remove that and I will open a new issue to restrict the keys to specific values

File: goldener/clusterize.py [hunk 15]
```diff
@@ -683,6 +750,20 @@ def _add_rows_to_cluster_table_from_dataset(
...
+                    for idx in batch[restriction_idx_key]
+                } - restrict_to
+                if to_remove:
+                    batch = filter_batch_from_indices(
+                        batch,
+                        to_remove,
+                        index_key=restriction_idx_key,
+                    )
+                    if len(batch) == 0:
+                        continue
```
[the remark this answers is not in the document]
Comment (Yann-CV) [reply goldener/clusterize.py:763]: we will fix that in another issue

**Review by Yann-CV — APPROVED** [review]
LGTM Good job!

Thanks again for your contribution!

Hope to review you again soon

**Comment by Yann-CV** [comment]
@TaniaW777 I added you as contributor here! you can now assign yourself to issues (security against duplicated work)

**Comment by Yann-CV** [comment]
@Mr-Neutr0n I added you as contributor here! you can now assign yourself to issues (security against duplicated work)

**Comment by Yann-CV** [comment]
> @TaniaW777 I added you as contributor here! you can now assign yourself to issues (security against duplicated work)

@TaniaW777 sorry I selected the wrong PR (this one is not from you) but at least you know it

**Comment by Mr-Neutr0n** [comment]
> I added you as contributor here! you can now assign yourself to issues (security against duplicated work)

Will do!
